File: scripts/status_logic.py
"""Shared status-badge derivation logic.

Mirrors the auto-calculated Status column in Lab_Papers_Dashboard.xlsx (see
the workbook's "Legend & Color Key" sheet for the source rules).
export_data.py does NOT use this — it reads the spreadsheet's own
precomputed Status text directly, since Excel already ran the formula.

This module exists for updates applied via the GitHub Issue Form / Action
pipeline (apply_update.py), where there is no spreadsheet formula to
consult, so the badge has to be computed from the raw Stage / Attempts /
Decision / Deadline / Published fields instead. The output strings must
stay byte-for-byte identical to the ones the spreadsheet produces, because
index.html's badge coloring and the "Needs Status" grouping both key off
substring matches (e.g. `status.indexOf("Overdue")`).
"""
from datetime import date

WAITING_STAGES = {"Idea", "On Hold", "Withdrawn"}

STATUS_PUBLISHED = "⚪ Published"
STATUS_COMPLETED = "⚪ Completed"
STATUS_CLOSED = "\U0001F534 Closed — resubmit?"
STATUS_NEEDS_ATTENTION = "\U0001F7E1 Needs Attention"
STATUS_WAITING = "\U0001F7E0 Waiting"
STATUS_OVERDUE = "\U0001F534 Overdue"
STATUS_REVIEWING = "\U0001F535 Reviewing"
STATUS_ON_TRACK = "\U0001F7E2 On Track"
STATUS_NEEDS_STATUS = "⚪ Needs Status"


def days_left(deadline_iso, today=None):
    """Days between `today` and an ISO date string (negative = past)."""
    if not deadline_iso:
        return None
    today = today or date.today()
    d = date.fromisoformat(deadline_iso[:10])
    return (d - today).days


def derive_status(stage, attempts, latest_decision, deadline_iso, published_date, today=None):
    """Recompute the Status badge from raw fields, per the workbook's rules.

    Order matters — these branches mirror the priority a human reading the
    Legend sheet top-to-bottom would apply (most-resolved state first).
    """
    if published_date:
        return STATUS_PUBLISHED
    if latest_decision == "Accepted":
        return STATUS_COMPLETED
    if latest_decision == "Rejected":
        return STATUS_CLOSED

    dleft = days_left(deadline_iso, today)

    if latest_decision == "Revise & Resubmit" or (dleft is not None and 0 <= dleft <= 14):
        return STATUS_NEEDS_ATTENTION

    if dleft is not None and dleft < 0:
        return STATUS_OVERDUE

    if stage in WAITING_STAGES:
        return STATUS_WAITING

    if (attempts or 0) > 0 or stage == "Internal Review":
        return STATUS_REVIEWING

    if stage == "Drafting":
        return STATUS_ON_TRACK

    return STATUS_NEEDS_STATUS

File: worker/src/status_logic.js
// Ported from scripts/status_logic.py — Cloudflare Workers run pure V8
// JavaScript (no subprocess, no Python runtime available inside a request),
// so "call out to the existing Python" isn't an option here; this is a
// deliberate line-for-line port, not a reimplementation from scratch.
//
// KEEP THIS IN SYNC with scripts/status_logic.py. worker/test/run.js cross-
// validates both against the same fixtures on every offline test run
// specifically to catch drift between the two — if you change the rules in
// one, the test will fail until you change the other too.
//
// Output strings must stay byte-for-byte identical to the Python version's,
// because index.html's badge coloring and "Needs Status" grouping both key
// off substring matches (e.g. `status.indexOf("Overdue")`).

const WAITING_STAGES = new Set(["Idea", "On Hold", "Withdrawn"]);

export const STATUS_PUBLISHED = "⚪ Published";
export const STATUS_COMPLETED = "⚪ Completed";
export const STATUS_CLOSED = "\u{1F534} Closed — resubmit?";
export const STATUS_NEEDS_ATTENTION = "\u{1F7E1} Needs Attention";
export const STATUS_WAITING = "\u{1F7E0} Waiting";
export const STATUS_OVERDUE = "\u{1F534} Overdue";
export const STATUS_REVIEWING = "\u{1F535} Reviewing";
export const STATUS_ON_TRACK = "\u{1F7E2} On Track";
export const STATUS_NEEDS_STATUS = "⚪ Needs Status";

/** Days between `today` and an ISO date string (negative = past). Mirrors
 * status_logic.py's days_left(). */
export function daysLeft(deadlineIso, today) {
  if (!deadlineIso) return null;
  const t = today || new Date();
  const todayUtc = Date.UTC(t.getUTCFullYear(), t.getUTCMonth(), t.getUTCDate());
  const d = new Date(deadlineIso.slice(0, 10) + "T00:00:00Z");
  const deadlineUtc = Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate());
  return Math.round((deadlineUtc - todayUtc) / 86400000);
}

/** Recompute the Status badge from raw fields. Mirrors status_logic.py's
 * derive_status() branch-for-branch, including the priority order. */
export function deriveStatus(stage, attempts, latestDecision, deadlineIso, publishedDate, today) {
  if (publishedDate) return STATUS_PUBLISHED;
  if (latestDecision === "Accepted") return STATUS_COMPLETED;
  if (latestDecision === "Rejected") return STATUS_CLOSED;

  const dleft = daysLeft(deadlineIso, today);

  if (latestDecision === "Revise & Resubmit" || (dleft !== null && dleft >= 0 && dleft <= 14)) {
    return STATUS_NEEDS_ATTENTION;
  }

  if (dleft !== null && dleft < 0) return STATUS_OVERDUE;

  if (WAITING_STAGES.has(stage)) return STATUS_WAITING;

  if ((attempts || 0) > 0 || stage === "Internal Review") return STATUS_REVIEWING;

  if (stage === "Drafting") return STATUS_ON_TRACK;

  return STATUS_NEEDS_STATUS;
}
